worker/auditconfigupdater/manifold.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package auditconfigupdater
+
+import (
+	"github.com/juju/errors"
+	"gopkg.in/juju/worker.v1"
+
+	jujuagent "github.com/juju/juju/agent"
+	"github.com/juju/juju/core/auditlog"
+	"github.com/juju/juju/worker/common"
+	"github.com/juju/juju/worker/dependency"
+	workerstate "github.com/juju/juju/worker/state"
+)
+
+// ManifoldConfig holds the information needed to run an
+// auditconfigupdater in a dependency.Engine.
+type ManifoldConfig struct {
+	AgentName string
+	StateName string
+	NewWorker func(ConfigSource, auditlog.Config, AuditLogFactory, chan<- auditlog.Config) (worker.Worker, error)
+}
+
+// Validate validates the manifold configuration.
+func (config ManifoldConfig) Validate() error {
+	if config.AgentName == "" {
+		return errors.NotValidf("empty AgentName")
+	}
+	if config.StateName == "" {
+		return errors.NotValidf("empty StateName")
+	}
+	if config.NewWorker == nil {
+		return errors.NotValidf("nil NewWorker")
+	}
+	return nil
+}
+
+// Manifold returns a dependency.Manifold to run an
+// auditconfigupdater.
+func Manifold(config ManifoldConfig) dependency.Manifold {
+	md := &manifoldData{
+		cfg:     config,
+		changes: make(chan auditlog.Config, 1),
+	}
+	return dependency.Manifold{
+		Inputs: []string{
+			config.AgentName,
+			config.StateName,
+		},
+		Start:  md.start,
+		Output: md.getOutput,
+	}
+}
+
+// Output defines what values the auditconfigupdater provides to workers that
+// depend on it.
+type Output interface {
+	Config() auditlog.Config
+	Changes() <-chan auditlog.Config
+}
+
+// manifoldData holds values needed for running the manifold and
+// providing outputs. It would have been called manifoldState if state
+// wasn't severely overused.
+type manifoldData struct {
+	cfg         ManifoldConfig
+	auditConfig auditlog.Config
+	changes     chan auditlog.Config
+}
+
+func (md *manifoldData) start(context dependency.Context) (worker.Worker, error) {
+	if err := md.cfg.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var agent jujuagent.Agent
+	if err := context.Get(md.cfg.AgentName, &agent); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var stTracker workerstate.StateTracker
+	if err := context.Get(md.cfg.StateName, &stTracker); err != nil {
+		return nil, errors.Trace(err)
+	}
+	statePool, err := stTracker.Use()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	logDir := agent.CurrentConfig().LogDir()
+
+	st := statePool.SystemState()
+
+	logFactory := func(cfg auditlog.Config) auditlog.AuditLog {
+		return auditlog.NewLogFile(logDir, cfg.MaxSizeMB, cfg.MaxBackups)
+	}
+	md.auditConfig, err = initialConfig(st)
+	if err != nil {
+		stTracker.Done()
```

Comment:
we should instead add a defer to the top of this function - remembering to call Done() everywhere is quite error prone. of course, this requires a named err return argument
defer func() {
  if err != nil && stTracker != nil {
    stTracker.Done()
  }
}

Reply:
:ok: